Allow solution of eigen problems with constraints

lindsayad opened the pull request:
Closes #27262. Prerequisite libMesh PR at https://github.com/libMesh/libmesh/pull/3821

lindsayad committed:
Allow solution of eigen problems with constraints

Closes #27262

lindsayad committed:
Add hanging node test

lindsayad committed:
Working on support for PJFNKMO

lindsayad commented on test/tests/problems/eigen_problem/eigensolvers/tests:
```diff
@@ -71,6 +71,15 @@
+    csvdiff = 'pjfnk_hanging_nodes_eigenvalues_0001.csv'
+    slepc_version = '>=3.8.0'
+    requirement = 'The system shall be able to solve nonlinear eigenvalue problems with hanging node constraints.'
+  []
```
👍

lindsayad committed:
Use get_subvector and restore_subvector for initial space

This allows us to use local dofs

lindsayad committed:
Add PJFNKMO test

lindsayad committed:
Initialize condensed matrices directly in MOOSE

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2383737/) on 7521a5e wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>a062c2</th>
      <th colspan=4>#27276 7521a5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.04%</td>
      <td>85.05%</td>
      <td>+0.02%</td>
      <td>93.48%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104966</td>
      <td>105034</td>
      <td>+68</td>
      <td>129</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18471</td>
      <td>18460</td>
      <td>-11</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27276/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27276/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27276/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented:
this will need to wait on libmesh/libmesh#3924 in order to pass the dbg targets, but I'm going to open this for review anyway

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2398663/) on 1071e69 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27276/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Apptainer moose-mpich](https://civet.inl.gov/job/2398635/) on 1071e69 : invalidated by @lindsayad

moosebuild commented:
Job [Apptainer moose-openmpi](https://civet.inl.gov/job/2398636/) on 1071e69 : invalidated by @lindsayad

moosebuild commented:
Job [Docker moose-dev](https://civet.inl.gov/job/2398644/) on 1071e69 : invalidated by @lindsayad

YaqiWang commented on framework/src/systems/NonlinearEigenSystem.C:
```diff
@@ -188,16 +190,54 @@ NonlinearEigenSystem::postAddResidualObject(ResidualObject & object)
   if (!presolve_succeeded)
     return;
 
+  std::unique_ptr<NumericVector<Number>> subvec;
```
We are creating a temporary vector here and cannot use the existing solution vector in place?

lindsayad replied:
The existing vector may be used in-place. See https://petsc.org/main/manualpages/Vec/VecGetSubVector/

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -549,73 +550,145 @@ mooseEPSFormMatrices(EigenProblem & eigen_problem, EPS eps, Vec x, void * ctx)
+  if (dof_map.n_constrained_dofs())
+    for (const auto i : index_range(mats))
+    {
+      PetscMatrix<Number> sub(mats[i], sys.comm());
```
Where is this sub used?

lindsayad replied:
the sub-matrices (the matrices representing only unconstrained dofs) are the matrices used in the eigensolves in SLEPc

YaqiWang commented on framework/include/utils/SlepcSupport.h:
```diff
@@ -80,13 +80,18 @@ void clearFreeNonlinearPowerIterations(const InputParameters & params);
 
 /**
  * Form multiple matrices for multiple tags
  */
```
It will be nice to add more doxygen comments, like using `@param` for these functions.

lindsayad commented:
@pbalest FYI we are close!

lindsayad committed:
Add some doxygen and hopefullying clarifying variable names

roystgnr commented on framework/include/utils/SlepcSupport.h:
```diff
@@ -80,13 +80,18 @@ void clearFreeNonlinearPowerIterations(const InputParameters & params);
  */
-void moosePetscSNESFormMatrixTag(SNES snes, Vec x, Mat mat, void * ctx, TagID tag);
+void moosePetscSNESFormMatrixTag(
+    SNES snes, Vec x, Mat mat, SparseMatrix<Number> & libmesh_mat, void * ctx, TagID tag);
```
The code later makes things clear, but we should still document the arguments' meanings here.

lindsayad replied:
Did you see the doxygen I pushed for @YaqiWang ?

lindsayad replied:
(and tweak of the variable's names)

roystgnr replied:
Looks like I was reviewing the version before that commit; that looks great.

roystgnr reviewed (changes requested):
A couple suggestions for possible optimization and future refactoring, but I think the only thing I really want to insist on is clear documentation for those function signatures.

lindsayad committed:
Rename condenseOutConstraints -> intializeCondensedMatrices

lindsayad commented on framework/src/systems/NonlinearEigenSystem.C:
```diff
@@ -188,16 +190,54 @@ NonlinearEigenSystem::postAddResidualObject(ResidualObject & object)
 }
 
+void
+NonlinearEigenSystem::condenseOutConstraints()
```
It would make sense since there is nothing MOOSEy in the method implementation

lindsayad replied:
I believe the reason I did not is because there has been no use case in libMesh yet, neither in the library nor in the example I've drummed up

lindsayad replied:
Ok I've opened https://github.com/libMesh/libmesh/pull/3936 but I would prefer to get this merged and then update the code in MOOSE later when we get around to another libmesh update

lindsayad replied:
Future MOOSE refactor at https://github.com/idaholab/moose/pull/28452/commits/bac300fcfc2b439a8f70c9ac301e5aaf374ffd08

lindsayad committed:
Only init matrices when needed

Now we should only call our constrained data initialization function at times
when the number of constrained dofs may have changed, e.g. during problem init
and if the mesh has changed
